go/vt/vtgate/planbuilder/select2.go

```diff
@@ -0,0 +1,367 @@
+		case *JoinBuilder:
+			return makeJoinBuilder(lplanBuilder, lsymbols, rplanBuilder, rsymbols, join)
+		case *RouteBuilder:
+			return joinRoutes(lplanBuilder, lsymbols, rplanBuilder, rsymbols, join)
+		}
+	}
+	panic("unreachable")
+}
+
+// makeJoinBuilder creates a new JoinBuilder node out of the two builders.
+// This function is called when the two builders cannot be part of
+// the same route.
+func makeJoinBuilder(lplanBuilder PlanBuilder, lsymbols *SymbolTable, rplanBuilder PlanBuilder, rsymbols *SymbolTable, join *sqlparser.JoinTableExpr) (PlanBuilder, *SymbolTable, error) {
+	err := lsymbols.Add(rsymbols)
+	if err != nil {
+		return nil, nil, err
+	}
+	isLeft := false
+	if join.Join == sqlparser.LeftJoinStr {
+		isLeft = true
+	}
+	assignOrder(rplanBuilder, lplanBuilder.Order())
+	return &JoinBuilder{
+		IsLeft: isLeft,
+		order:  rplanBuilder.Order(),
+		Left:   lplanBuilder,
+		Right:  rplanBuilder,
+	}, lsymbols, nil
+}
+
+// assignOrder sets the order for the nodes of the tree based on the
+// starting order.
+func assignOrder(planBuilder PlanBuilder, order int) {
+	switch planBuilder := planBuilder.(type) {
+	case *JoinBuilder:
+		assignOrder(planBuilder.Left, order)
+		assignOrder(planBuilder.Right, planBuilder.Left.Order())
+		planBuilder.order = planBuilder.Right.Order()
+	case *RouteBuilder:
+		planBuilder.order = order + 1
+	}
+}
+
+// joinRoutes attempts to join two RouteBuilder objects into one.
+// If it's possible, it produces a joined RouteBuilder.
+// Otherwise, it's a JoinBuilder.
+func joinRoutes(lRouteBuilder *RouteBuilder, lsymbols *SymbolTable, rRouteBuilder *RouteBuilder, rsymbols *SymbolTable, join *sqlparser.JoinTableExpr) (PlanBuilder, *SymbolTable, error) {
+	if lRouteBuilder.Route.Keyspace.Name != rRouteBuilder.Route.Keyspace.Name {
+		return makeJoinBuilder(lRouteBuilder, lsymbols, rRouteBuilder, rsymbols, join)
+	}
+	if lRouteBuilder.Route.PlanID == SelectUnsharded {
+		if rRouteBuilder.Route.PlanID == SelectUnsharded {
+			// Two Routes from the same unsharded keyspace can be merged.
+			return mergeRoutes(lRouteBuilder, lsymbols, rsymbols, join)
+		}
+		return makeJoinBuilder(lRouteBuilder, lsymbols, rRouteBuilder, rsymbols, join)
+	}
+	// lRouteBuilder is a sharded route. It can't merge with an unsharded route.
+	if rRouteBuilder.Route.PlanID == SelectUnsharded {
+		return makeJoinBuilder(lRouteBuilder, lsymbols, rRouteBuilder, rsymbols, join)
+	}
+	// TODO(sougou): Handle special case for SelectEqual and SelectKeyrange.
+	// Both RouteBuilder are sharded routes. Analyze join condition for merging.
+	return joinShardedRoutes(lRouteBuilder, lsymbols, rRouteBuilder, rsymbols, join)
+}
+
+// mergeRoutes makes a new RouteBuilder by joining the left and right
+// nodes of a join. This is called if two routes can be merged.
+func mergeRoutes(lRouteBuilder *RouteBuilder, lsymbols, rsymbols *SymbolTable, join *sqlparser.JoinTableExpr) (PlanBuilder, *SymbolTable, error) {
```

Comment:
The more correct way of expressing it is that we're creating a new merged route that inherits the routing properties of the left node. The implementation reuses the left route as an optimization.
Looks like I'm missing the comment about the left node. I'll add that.